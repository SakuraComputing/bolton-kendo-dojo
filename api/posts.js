const express = require('express');
const router = express.Router();
const mongoose = require('mongoose');
const passport = require('passport');


// Post model
const Post = require('./models/Post');

// Va;idation
const validatePostInput = require('../validation/post');

//@route GET api/posts/test
// @desc Test post route
// @access Private
router.get('/test', (req, res) => res.json({
    mmsg: 'Posts works'
}));

// @route GET api/posts
// @desc Get posts
// @access Public
router.get('/', (req, res) => {
    Post.find()
    .sort({ date: -1 })
    .then(posts => res.json(posts))
    .catch(err => res.status(404).json({ nopostsfound: "No posts found"}));
})

// @route GET api/posts/:id
// @desc Get post by id
// @access Public
router.get('/:id', (req, res) => {
    Post.findById(req.params.id)
    .then(post => res.json(post))
    .catch(err => res.status(404).json({ nopostfound: "No post found with that Id"}));
})


// @route POST api/posts
// @desc Create post route
// @access Private
router.post('/', passport.authenticate('jwt', { session: false }), (req, res) => {

    const { errors, isValid } = validatePostInput(req.body);

    // Check validation
    if(!isValid) {
        return res.status(400).json(errors);
    }

    console.log(req.body.text);

    const newPost = new Post({
        text: req.body.text,
        name: req.body.name,
        avatar: req.body.avatar, 
        user: req.user.id
    });

    newPost.save().then(post => res.json(post))
});

// @route DELETE api/posts/:id
// @desc Delete post by Id
// @access Private
router.delete('/:id', passport.authenticate('jwt', { session: false }), (req, res) => {
    Post.findById(req.params.id)
    .then(post => {
        if(post.user.toString() !== req.user.id) {
            return res.status(401).json({ notauthorised: "User not authorized "})
        }
        post.remove().then(() => res.json({ success: true }))
        .catch(err => res.statusMessage(404).json({ postnotfound: "No post found "}))
    })
})


module.exports = router;