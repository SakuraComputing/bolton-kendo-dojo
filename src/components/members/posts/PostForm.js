import React, { Component } from 'react';
import { connect } from 'react-redux';
import TextAreaFieldGroup from '../../../common/TextAreaFieldGroup';
import { addPost } from '../../../actions/postActions';

export class PostForm extends Component {

    constructor(props) {
        super(props)
        this.state = {
            text: '',
            errors: {}
        }
    }

    onChange = (e) => {
        this.setState({ text: e.target.value })
    }
    
    render() {

        const { errors } = this.state;

        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <div className="form-group">
                                <TextAreaFieldGroup
                                    className="post-text"
                                    placeholder="Create a post"
                                    name="text"
                                    value={this.state.text}
                                    onChange={this.onChange}
                                    error={errors.text}
                                />
                            </div>
                    <button type="submit" className="post-submit">
                        Submit
                    </button>                
                </form>
            </div>
        )
    }
}
const mapStateToProps = state => ({
    auth: state.auth,
    errors: state.errors
})
export default connect(mapStateToProps, { addPost })(PostForm);
