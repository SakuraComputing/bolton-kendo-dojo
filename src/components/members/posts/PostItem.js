import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { deletePost } from "../../../actions/postActions";

export class PostItem extends Component {

    onDeleteClick = (id) => {
        this.props.deletePost(id);
    }
    render() {

        const { post } = this.props;

        return (
            <div className="feed-container">
                <div>
                    <p>Posted By: {post.name}</p>
                </div>
                <div>
                    <p>{post.text}</p>
                    <Link to={`/post/${post._id}`} className="item-comments button-small">
                        Comments
                    </Link>
                    <button 
                        type="button"
                        className="button-delete button-small item-delete"
                        onClick={this.onDeleteClick.bind(this, post._id)}
                    >
                    Delete Post
                    </button>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    auth: state.auth
});

export default connect(mapStateToProps, { deletePost })(PostItem);
