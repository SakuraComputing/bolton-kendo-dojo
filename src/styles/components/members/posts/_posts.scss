.feed {
    margin: 0 2%;
}

.post-submit {
    margin-left: 10px;
}
.feed-container {
    margin: 0px 20px;
    border: solid 1px $blue-black;
    border-radius: 5px;
    padding: 2%;
    font-size: 110%;
}

.item-delete {
    float: right;
    margin-top: 0;
}