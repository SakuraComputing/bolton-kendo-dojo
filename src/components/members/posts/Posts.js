import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getPosts } from '../../../actions/postActions';
import PostForm from './PostForm';
import Spinner from '../../../common/Spinner';
import PostFeed from './PostFeed';

export class Posts extends Component {

  componentDidMount() {
      this.props.getPosts();
  }
  
  render() {

    const { posts, loading } = this.props.post;

    let postContent;

    if(posts == null || posts === undefined || loading) {
        postContent = <Spinner />
    } else {
        postContent = <PostFeed posts={posts} />
    }

    return (
      <div>
        <PostForm />
        {postContent}
      </div>
    )
  }
}

const mapStateToProps = state => ({
    post: state.post
})
export default connect(mapStateToProps, { getPosts })(Posts);
