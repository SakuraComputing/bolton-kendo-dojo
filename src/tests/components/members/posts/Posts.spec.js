import React from 'react';
import { shallow } from 'enzyme';
import { Posts } from '../../../../components/members/posts/Posts';

describe('Posts Component', () => {
  let posts;

  beforeEach(() => {
      const props = {
          post: {
              posts: [],
              loading: false
          }
      }
      posts = shallow(<Posts {...props}/>)
  });
  it('should render correctly', () => {
      expect(posts).toMatchSnapshot();
  });
  it('should contain a PostForm component', () => {
      expect(posts.find('Connect(PostForm)').exists()).toBe(true);
  });
})
