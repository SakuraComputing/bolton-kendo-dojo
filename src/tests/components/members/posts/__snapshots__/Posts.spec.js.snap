// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts Component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Posts
    getPosts={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    post={
      Object {
        "loading": false,
        "posts": Array [],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(PostForm) />,
        <PostFeed
          posts={Array []}
        />,
      ],
      "className": "feed",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "posts": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(PostForm) />,
          <PostFeed
            posts={Array []}
          />,
        ],
        "className": "feed",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "posts": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Posts Component should render the spinner component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Posts
    getPosts={
      [MockFunction] {
        "calls": Array [
          Array [],
          Array [],
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    post={
      Object {
        "loading": false,
        "posts": Array [],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(PostForm) />,
        <PostFeed
          posts={Array []}
        />,
      ],
      "className": "feed",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "posts": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(PostForm) />,
          <PostFeed
            posts={Array []}
          />,
        ],
        "className": "feed",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "posts": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Posts Component when mounted should dispatch the \`getPosts\` method it receives from props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Posts
    getPosts={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    post={
      Object {
        "loading": false,
        "posts": Array [],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(PostForm) />,
        <PostFeed
          posts={Array []}
        />,
      ],
      "className": "feed",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "posts": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(PostForm) />,
          <PostFeed
            posts={Array []}
          />,
        ],
        "className": "feed",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "posts": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
